core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/ArbitraryDistributionSplitAssigner.java

```diff
@@ -341,30 +341,44 @@ public String toString()
                 .toString();
     }
 
-    private Optional<HostAddress> getHostRequirement(Split split)
+    /**
+     * Ranks the desirability of selecting a node for the next split assignment.  Lower is better.
+     */
+    private long rank(HostAddress address)
     {
-        if (split.getConnectorSplit().isRemotelyAccessible()) {
-            return Optional.empty();
+        // The node-to-split map can have two entries for this address: one for remotely accessible splits and one for non remotely accessible splits.
+        PartitionAssignment flexEntry = openAssignments.get(new NodeRequirements(catalogRequirement, ImmutableSet.of(address), true));
+        PartitionAssignment rigidEntry = openAssignments.get(new NodeRequirements(catalogRequirement, ImmutableSet.of(address), false));
+        if (flexEntry == null && rigidEntry == null) {
+            return -1; // Most desirable: an unassigned node.
+        }
+        // The more data assigned to this node, the less desirable it is.
+        if (flexEntry == null) {
+            return rigidEntry.getAssignedDataSizeInBytes();
+        }
+        if (rigidEntry == null) {
+            return flexEntry.getAssignedDataSizeInBytes();
+        }
+        return flexEntry.getAssignedDataSizeInBytes() + rigidEntry.getAssignedDataSizeInBytes();
+    }
+
+    private NodeRequirements getNodeRequirements(Split split)
+    {
+        if (split.getAddresses().isEmpty()) {
+            return new NodeRequirements(catalogRequirement, ImmutableSet.of(), true);
         }
         List<HostAddress> addresses = split.getAddresses();
-        checkArgument(!addresses.isEmpty(), "split is not remotely accessible but the list of hosts is empty: %s", split);
```

Comment:
It's a pity to lose this check; you can slot it neatly two lines up (inside `if split.addresses.empty`).

Reply:
I'm not sure what you mean. I don't need `!addresses.isEmpty()` check under `if (split.getAddresses().isEmpty()) {` condition

Reply:
You would check `isRemotelyAccessible()` there, to achieve the same condition checked here, but in different order.

Reply:
you are right, I will change it.